### PR TITLE
ref(config): override config from URL on mobile

### PR DESCRIPTION
**Pull request by paweldomas**
Moves the things around to be able to override the config with the URL
params specified in the hash part of the location URI to which the app
is navigating to.

### CODE REVIEW

**Comment by paweldomas** [comment]
The reason for failing tests was that the PR test machine has it's own config.js (but it should come from the code) and it had the old p2pEnabled: true, which was conflicting with what we set through the URL (config.p2p.enabled=false). Because the two properties were present at the same time the old one was not overwritten correctly (not sure if it's a bug or expected behaviour).

Now I expect that the test will succeed, because the property was changed to the new style on the PR machine.

**Comment by damencho** [comment]
This is a valid situation that can happen on client deployments. I think, that if we introduce a new property which should override and deprecate an old one, it should take care of its values in the code when tha values of the new one changes ... or maybe not as this comes as URL params... Or at least ignore the old one if there are values for the new one.

**Comment by paweldomas** [comment]
The situation is not expected when running PR tests. PR machine having predefined config was the last thing I was expecting since the app is built from the branch and it took me 2 days to dig into that, but I should have known better what to expect.

**Comment by damencho** [comment]
Hum... It is normal deployment as any other client would deploy it ... Just the config files are in the home folder of the user running the tests.

**Comment by paweldomas** [comment]
But the config does not come from source, does it ? It is edited by hand.

**Comment by damencho** [comment]
Yep, as any other deployment. The config comes from a template from source at the time of deployment and edited by the user, and can stay forever like that.

**Comment by paweldomas** [comment]
So there are 2 sides of the story:
1. whether or not the legacy handling code is good enough - it could be improved
2. should the PR tests have static config - the PR tests machine IMHO should get the config from source. It's not like "any other deployment" - it's a PR testing deployment that tests particular branch of the code which should include config.js contents of that branch since we're treating is as part of the source.

**Comment by damencho** [comment]
2. But this way you will skip all cases like this. You will notice them after we build packages, move them to stable and people start complaining. 
Config is always static and when you deploy its static. We warn user for cahnges, but when one property goes in it we should take care of its value.
If we are going to deprecate and will start ignoring it, the minimum we are doing till now is sending a mail to users and dev mailinglist.
If we are generating it on every test, this will be the only usecase of this kind of deployment ... then when it goes to beta it can brake.

**Comment by paweldomas** [comment]
Testing of the legacy properties is another story, which eventually would require a new test case or unit tests. It should not cause the confusion during PR tests.

**Comment by damencho** [comment]
The code should not depend on configuration. It should work with defaults, if you need to test something special write a test case.

**Comment by paweldomas** [comment]
legacy config is a special case

**Comment by paweldomas** [comment]
You're claiming that having the static config on PR machine tests something, but now I changed the config and it's no longer being tested.

**Comment by paweldomas** [comment]
It was "testing by coincidence" or more like a debugging trap, because it's not expected.

**Comment by paweldomas** [comment]
Anyway the PR is not reviewed yet, so it's not said I won't be fixing the legacy code handling.

**Comment by lyubomir** [comment]
@paweldomas and @damencho, I just read the conversation above about `p2pEnabled` and the legacy support but I'm afraid I don't understand whether it's a concern that should be blocking this PR. If I'm not mistaken, we wanted to parse the URL params into config on mobile here which already existed on Web and the question about conflicting property values, legacy property support, etc. is not really related to a platform/OS, right?

On the legacy support, we did have a function to do legacy to latest format translation (and if I'm not mistaken at some point it was about p2p specifically). I understand in this case it's not that but a config with conflicting values which the app can always stumble upon anyway with a user-modified config.js. But I don't see right now any pending work in that regard as blocking this PR.

File: react/features/base/config/reducer.js [hunk 1]
```diff
@@ -61,8 +61,21 @@ ReducerRegistry.register(
...
+            const newConfig = _setConfig(state, action);
+
+            // FIXME On web we rely on the global 'config' variable which gets
+            // altered multiple times, before it makes it to the reducer. At
+            // some point it may not be the global variable which is being
+            // modified anymore due to different merge methods being used
+            // across the way. The global variable must be synchronized with
+            // the final state resolved by the reducer.
+            if (typeof window !== 'undefined' && window.config) {
+                window.config = newConfig;
```
Comment (lyubomir) [inline react/features/base/config/reducer.js]: The assignment to `window.config` is a side effect and, consequently, doesn't belong in a reducer. Was there a problem when performing the assignment in a middleware?

File: react/features/app/actions.js [hunk 2]
```diff
@@ -1,7 +1,11 @@
 /* @flow */
 
 import { setRoom } from '../base/conference';
-import { configWillLoad, loadConfigError, setConfig } from '../base/config';
+import {
+    configWillLoad,
+    loadConfigError,
+    setConfig
+} from '../base/config';
```
Comment (lyubomir) [inline react/features/app/actions.js]: Is that a coding style/formatting change or is it a leftover from earlier versions of this PR?

File: react/features/base/config/reducer.js (hunk 1 above)
Comment (paweldomas) [reply react/features/base/config/reducer.js]: yup, stupid me

**Comment by paweldomas** [comment]
not tested yet - do not merge

**Pull request merged**